[FEAT] Delete user token and associated notifications on DEVICE UNREGISTER event

**Is your feature request related to a problem? Please describe.**
There are several errors when sending notifications because the devices could be unregistered and there is no handling of this error code.

**Describe the solution you'd like**
When a notification is sent to a device that is unregistered, FCM SDK reports its error code. This can be used to delete the FCM token for the user and also delete any scheduled notifications for them

The change can be added here https://github.com/RADAR-base/RADAR-Appserver/blob/f4201d482641588d7ace390aab72baaa8d306694/src/main/java/org/radarbase/appserver/service/scheduler/MessageSchedulerService.java#L237

will also require adding a parameter of type `Message` to ` handleFCMErrorCode` so can get details on the user from https://github.com/RADAR-base/RADAR-Appserver/blob/f4201d482641588d7ace390aab72baaa8d306694/src/main/java/org/radarbase/appserver/service/scheduler/NotificationSchedulerService.java#L102.

**Describe alternatives you've considered**
A clear and concise description of any alternative solutions or features you've considered.

**Priority**
2

**Difficulty**
3

**Additional context**
Relevant Logs-
```
2021-06-25 17:30:00,112 ERROR [quartzScheduler_Worker-3] org.radarbase.appserver.service.scheduler.NotificationSchedulerService: Error occurred when sending downstream message.
com.google.firebase.messaging.FirebaseMessagingException: Requested entity was not found.
	at com.google.firebase.messaging.FirebaseMessagingException.withMessagingErrorCode(FirebaseMessagingException.java:47)
	at com.google.firebase.messaging.FirebaseMessagingClientImpl$MessagingErrorHandler.createException(FirebaseMessagingClientImpl.java:293)
	at com.google.firebase.messaging.FirebaseMessagingClientImpl$MessagingErrorHandler.createException(FirebaseMessagingClientImpl.java:282)
	at com.google.firebase.internal.AbstractHttpErrorHandler.handleHttpResponseException(AbstractHttpErrorHandler.java:57)
	at com.google.firebase.internal.ErrorHandlingHttpClient.send(ErrorHandlingHttpClient.java:108)
	at com.google.firebase.internal.ErrorHandlingHttpClient.sendAndParse(ErrorHandlingHttpClient.java:72)
	at com.google.firebase.messaging.FirebaseMessagingClientImpl.sendSingleRequest(FirebaseMessagingClientImpl.java:127)
	at com.google.firebase.messaging.FirebaseMessagingClientImpl.send(FirebaseMessagingClientImpl.java:113)
	at com.google.firebase.messaging.FirebaseMessaging$1.execute(FirebaseMessaging.java:135)
	at com.google.firebase.messaging.FirebaseMessaging$1.execute(FirebaseMessaging.java:132)
	at com.google.firebase.internal.CallableOperation.call(CallableOperation.java:36)
	at com.google.firebase.messaging.FirebaseMessaging.send(FirebaseMessaging.java:102)
	at com.google.firebase.messaging.FirebaseMessaging.send(FirebaseMessaging.java:86)
	at org.radarbase.fcm.downstream.AdminSdkFcmSender.send(AdminSdkFcmSender.java:117)
	at org.radarbase.appserver.service.scheduler.NotificationSchedulerService.send(NotificationSchedulerService.java:97)
	at org.radarbase.appserver.service.scheduler.quartz.MessageJob.execute(MessageJob.java:91)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: com.google.api.client.http.HttpResponseException: 404 Not Found
POST https://fcm.googleapis.com/v1/projects/radar-armt-notification/messages:send
{
  "error": {
    "code": 404,
    "message": "Requested entity was not found.",
    "status": "NOT_FOUND",
    "details": [
      {
        "@type": "type.googleapis.com/google.firebase.fcm.v1.FcmError",
        "errorCode": "UNREGISTERED"
      }
    ]
  }
}

	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1113)
	at com.google.firebase.internal.ErrorHandlingHttpClient.send(ErrorHandlingHttpClient.java:96)
	... 13 common frames omitted
```

--- a/src/main/java/org/radarbase/appserver/service/scheduler/DataMessageSchedulerService.java
+++ b/src/main/java/org/radarbase/appserver/service/scheduler/DataMessageSchedulerService.java
@@ -25,6 +25,9 @@ import com.google.firebase.messaging.FirebaseMessagingException;
 import java.util.Objects;
 import lombok.extern.slf4j.Slf4j;
 import org.radarbase.appserver.entity.DataMessage;
+import org.radarbase.appserver.service.FcmDataMessageService;
+import org.radarbase.appserver.service.FcmNotificationService;
+import org.radarbase.appserver.service.UserService;
 import org.radarbase.appserver.service.scheduler.quartz.SchedulerService;
 import org.radarbase.fcm.downstream.FcmSender;
 import org.radarbase.fcm.model.FcmDataMessage;
@@ -46,8 +49,11 @@ public class DataMessageSchedulerService extends MessageSchedulerService<DataMes
 
     public DataMessageSchedulerService(
             @Autowired @Qualifier("fcmSenderProps") FcmSender fcmSender,
-            @Autowired SchedulerService schedulerService) {
-        super(fcmSender, schedulerService);
+            @Autowired SchedulerService schedulerService,
+            @Autowired UserService userService,
+            @Autowired FcmNotificationService notificationService,
+            @Autowired FcmDataMessageService dataMessageService) {
+        super(fcmSender, schedulerService, userService, notificationService, dataMessageService);
     }
 
     private static FcmDataMessage createMessageFromDataMessage(DataMessage dataMessage) {
@@ -73,8 +79,8 @@ public class DataMessageSchedulerService extends MessageSchedulerService<DataMes
         } catch (FirebaseMessagingException exc) {
             log.error("Error occurred when sending downstream message.", exc);
             // TODO: update the data message status using event
-            handleErrorCode(exc.getErrorCode());
-            handleFCMErrorCode(exc.getMessagingErrorCode());
+            handleErrorCode(exc.getErrorCode(), dataMessage);
+            handleFCMErrorCode(exc.getMessagingErrorCode(), dataMessage);
         }
     }
 }

--- a/src/main/java/org/radarbase/appserver/service/scheduler/DataMessageSchedulerService.java
+++ b/src/main/java/org/radarbase/appserver/service/scheduler/DataMessageSchedulerService.java
@@ -21,13 +21,9 @@
 
 package org.radarbase.appserver.service.scheduler;
 
-import com.google.firebase.messaging.FirebaseMessagingException;
 import java.util.Objects;
 import lombok.extern.slf4j.Slf4j;
 import org.radarbase.appserver.entity.DataMessage;
-import org.radarbase.appserver.service.FcmDataMessageService;
-import org.radarbase.appserver.service.FcmNotificationService;
-import org.radarbase.appserver.service.UserService;
 import org.radarbase.appserver.service.scheduler.quartz.SchedulerService;
 import org.radarbase.fcm.downstream.FcmSender;
 import org.radarbase.fcm.model.FcmDataMessage;
@@ -49,11 +45,8 @@ public class DataMessageSchedulerService extends MessageSchedulerService<DataMes
 
     public DataMessageSchedulerService(
             @Autowired @Qualifier("fcmSenderProps") FcmSender fcmSender,
-            @Autowired SchedulerService schedulerService,
-            @Autowired UserService userService,
-            @Autowired FcmNotificationService notificationService,
-            @Autowired FcmDataMessageService dataMessageService) {
-        super(fcmSender, schedulerService, userService, notificationService, dataMessageService);
+            @Autowired SchedulerService schedulerService) {
+        super(fcmSender, schedulerService);
     }
 
     private static FcmDataMessage createMessageFromDataMessage(DataMessage dataMessage) {
@@ -74,13 +67,6 @@ public class DataMessageSchedulerService extends MessageSchedulerService<DataMes
     }
 
     public void send(DataMessage dataMessage) throws Exception {
-        try {
-            fcmSender.send(createMessageFromDataMessage(dataMessage));
-        } catch (FirebaseMessagingException exc) {
-            log.error("Error occurred when sending downstream message.", exc);
-            // TODO: update the data message status using event
-            handleErrorCode(exc.getErrorCode(), dataMessage);
-            handleFCMErrorCode(exc.getMessagingErrorCode(), dataMessage);
-        }
+        fcmSender.send(createMessageFromDataMessage(dataMessage));
     }
 }

--- a/src/main/java/org/radarbase/appserver/service/scheduler/MessageSchedulerService.java
+++ b/src/main/java/org/radarbase/appserver/service/scheduler/MessageSchedulerService.java
@@ -37,10 +37,14 @@ import org.quartz.JobKey;
 import org.quartz.SimpleTrigger;
 import org.quartz.Trigger;
 import org.quartz.TriggerKey;
+import org.radarbase.appserver.dto.fcm.FcmUserDto;
 import org.radarbase.appserver.entity.DataMessage;
 import org.radarbase.appserver.entity.Message;
 import org.radarbase.appserver.entity.Notification;
+import org.radarbase.appserver.service.FcmDataMessageService;
+import org.radarbase.appserver.service.FcmNotificationService;
 import org.radarbase.appserver.service.MessageType;
+import org.radarbase.appserver.service.UserService;
 import org.radarbase.appserver.service.scheduler.quartz.MessageJob;
 import org.radarbase.appserver.service.scheduler.quartz.QuartzNamingStrategy;
 import org.radarbase.appserver.service.scheduler.quartz.SchedulerService;
@@ -63,12 +67,21 @@ public abstract class MessageSchedulerService<T extends Message> {
     protected static final boolean IS_DELIVERY_RECEIPT_REQUESTED = true;
     protected final transient FcmSender fcmSender;
     protected final transient SchedulerService schedulerService;
+    protected final transient UserService userService;
+    protected final transient FcmNotificationService notificationService;
+    protected final transient FcmDataMessageService dataMessageService;
 
     public MessageSchedulerService(
             @Autowired @Qualifier("fcmSenderProps") FcmSender fcmSender,
-            @Autowired SchedulerService schedulerService) {
+            @Autowired SchedulerService schedulerService,
+            @Autowired UserService userService,
+            @Autowired FcmNotificationService notificationService,
+            @Autowired FcmDataMessageService dataMessageService) {
         this.fcmSender = fcmSender;
         this.schedulerService = schedulerService;
+        this.userService = userService;
+        this.notificationService = notificationService;
+        this.dataMessageService = dataMessageService;
     }
 
     public abstract void send(T message) throws Exception;
@@ -193,7 +206,7 @@ public abstract class MessageSchedulerService<T extends Message> {
         }
     }
 
-    protected void handleErrorCode(ErrorCode errorCode) {
+    protected void handleErrorCode(ErrorCode errorCode, Message message) {
         // More info on ErrorCode: https://firebase.google.com/docs/reference/fcm/rest/v1/ErrorCode
         switch (errorCode) {
             case INVALID_ARGUMENT:
@@ -220,7 +233,7 @@ public abstract class MessageSchedulerService<T extends Message> {
         }
     }
 
-    protected void handleFCMErrorCode(MessagingErrorCode errorCode) {
+    protected void handleFCMErrorCode(MessagingErrorCode errorCode, Message message) {
         switch (errorCode) {
             case INTERNAL:
             case QUOTA_EXCEEDED:
@@ -233,8 +246,13 @@ public abstract class MessageSchedulerService<T extends Message> {
                 log.warn("The FCM service is unavailable.");
                 break;
             case UNREGISTERED:
-                //TODO: remove all scheduled notifications/messages for this user.
-                log.warn("The Device was unregistered.");
+                FcmUserDto userDto = new FcmUserDto(message.getUser()).setFcmToken("undefined");
+                log.warn("The Device for user {} was unregistered.", userDto.getSubjectId());
+                notificationService
+                        .removeNotificationsForUser(userDto.getProjectId(), userDto.getSubjectId());
+                dataMessageService
+                        .removeDataMessagesForUser(userDto.getProjectId(), userDto.getSubjectId());
+                userService.updateUser(userDto);
                 break;
         }
     }

--- a/src/main/java/org/radarbase/appserver/service/scheduler/MessageSchedulerService.java
+++ b/src/main/java/org/radarbase/appserver/service/scheduler/MessageSchedulerService.java
@@ -21,8 +21,6 @@
 
 package org.radarbase.appserver.service.scheduler;
 
-import com.google.firebase.ErrorCode;
-import com.google.firebase.messaging.MessagingErrorCode;
 import java.util.Date;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -37,14 +35,10 @@ import org.quartz.JobKey;
 import org.quartz.SimpleTrigger;
 import org.quartz.Trigger;
 import org.quartz.TriggerKey;
-import org.radarbase.appserver.dto.fcm.FcmUserDto;
 import org.radarbase.appserver.entity.DataMessage;
 import org.radarbase.appserver.entity.Message;
 import org.radarbase.appserver.entity.Notification;
-import org.radarbase.appserver.service.FcmDataMessageService;
-import org.radarbase.appserver.service.FcmNotificationService;
 import org.radarbase.appserver.service.MessageType;
-import org.radarbase.appserver.service.UserService;
 import org.radarbase.appserver.service.scheduler.quartz.MessageJob;
 import org.radarbase.appserver.service.scheduler.quartz.QuartzNamingStrategy;
 import org.radarbase.appserver.service.scheduler.quartz.SchedulerService;
@@ -67,21 +61,12 @@ public abstract class MessageSchedulerService<T extends Message> {
     protected static final boolean IS_DELIVERY_RECEIPT_REQUESTED = true;
     protected final transient FcmSender fcmSender;
     protected final transient SchedulerService schedulerService;
-    protected final transient UserService userService;
-    protected final transient FcmNotificationService notificationService;
-    protected final transient FcmDataMessageService dataMessageService;
 
     public MessageSchedulerService(
             @Autowired @Qualifier("fcmSenderProps") FcmSender fcmSender,
-            @Autowired SchedulerService schedulerService,
-            @Autowired UserService userService,
-            @Autowired FcmNotificationService notificationService,
-            @Autowired FcmDataMessageService dataMessageService) {
+            @Autowired SchedulerService schedulerService) {
         this.fcmSender = fcmSender;
         this.schedulerService = schedulerService;
-        this.userService = userService;
-        this.notificationService = notificationService;
-        this.dataMessageService = dataMessageService;
     }
 
     public abstract void send(T message) throws Exception;
@@ -203,57 +188,6 @@ public abstract class MessageSchedulerService<T extends Message> {
             return MessageType.DATA;
         } else {
             return MessageType.UNKNOWN;
-        }
-    }
-
-    protected void handleErrorCode(ErrorCode errorCode, Message message) {
-        // More info on ErrorCode: https://firebase.google.com/docs/reference/fcm/rest/v1/ErrorCode
-        switch (errorCode) {
-            case INVALID_ARGUMENT:
-            case INTERNAL:
-            case ABORTED:
-            case CONFLICT:
-            case CANCELLED:
-            case DATA_LOSS:
-            case NOT_FOUND:
-            case OUT_OF_RANGE:
-            case ALREADY_EXISTS:
-            case DEADLINE_EXCEEDED:
-            case PERMISSION_DENIED:
-            case RESOURCE_EXHAUSTED:
-            case FAILED_PRECONDITION:
-            case UNAUTHENTICATED:
-            case UNKNOWN:
-                break;
-            case UNAVAILABLE:
-                // TODO: Could schedule for retry.
-                log.warn("The FCM service is unavailable.");
-                break;
-
-        }
-    }
-
-    protected void handleFCMErrorCode(MessagingErrorCode errorCode, Message message) {
-        switch (errorCode) {
-            case INTERNAL:
-            case QUOTA_EXCEEDED:
-            case INVALID_ARGUMENT:
-            case SENDER_ID_MISMATCH:
-            case THIRD_PARTY_AUTH_ERROR:
-                break;
-            case UNAVAILABLE:
-                // TODO: Could schedule for retry.
-                log.warn("The FCM service is unavailable.");
-                break;
-            case UNREGISTERED:
-                FcmUserDto userDto = new FcmUserDto(message.getUser()).setFcmToken("undefined");
-                log.warn("The Device for user {} was unregistered.", userDto.getSubjectId());
-                notificationService
-                        .removeNotificationsForUser(userDto.getProjectId(), userDto.getSubjectId());
-                dataMessageService
-                        .removeDataMessagesForUser(userDto.getProjectId(), userDto.getSubjectId());
-                userService.updateUser(userDto);
-                break;
         }
     }
 }

--- a/src/main/java/org/radarbase/appserver/service/scheduler/NotificationSchedulerService.java
+++ b/src/main/java/org/radarbase/appserver/service/scheduler/NotificationSchedulerService.java
@@ -21,15 +21,11 @@
 
 package org.radarbase.appserver.service.scheduler;
 
-import com.google.firebase.messaging.FirebaseMessagingException;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Objects;
 import lombok.extern.slf4j.Slf4j;
 import org.radarbase.appserver.entity.Notification;
-import org.radarbase.appserver.service.FcmDataMessageService;
-import org.radarbase.appserver.service.FcmNotificationService;
-import org.radarbase.appserver.service.UserService;
 import org.radarbase.appserver.service.scheduler.quartz.SchedulerService;
 import org.radarbase.fcm.downstream.FcmSender;
 import org.radarbase.fcm.model.FcmNotificationMessage;
@@ -50,11 +46,8 @@ public class NotificationSchedulerService extends MessageSchedulerService<Notifi
 
     public NotificationSchedulerService(
             @Autowired @Qualifier("fcmSenderProps") FcmSender fcmSender,
-            @Autowired SchedulerService schedulerService,
-            @Autowired UserService userService,
-            @Autowired FcmNotificationService notificationService,
-            @Autowired FcmDataMessageService dataMessageService) {
-        super(fcmSender, schedulerService, userService, notificationService, dataMessageService);
+            @Autowired SchedulerService schedulerService) {
+        super(fcmSender, schedulerService);
     }
 
     private static Map<String, Object> getNotificationMap(Notification notification) {
@@ -99,13 +92,6 @@ public class NotificationSchedulerService extends MessageSchedulerService<Notifi
 
 
     public void send(Notification notification) throws Exception {
-        try {
         fcmSender.send(createMessageFromNotification(notification));
-        } catch (FirebaseMessagingException exc) {
-            log.error("Error occurred when sending downstream message.", exc);
-            // TODO: update the notification status using event
-            handleErrorCode(exc.getErrorCode(), notification);
-            handleFCMErrorCode(exc.getMessagingErrorCode(), notification);
-        }
     }
 }

--- a/src/main/java/org/radarbase/appserver/service/scheduler/NotificationSchedulerService.java
+++ b/src/main/java/org/radarbase/appserver/service/scheduler/NotificationSchedulerService.java
@@ -27,6 +27,9 @@ import java.util.Map;
 import java.util.Objects;
 import lombok.extern.slf4j.Slf4j;
 import org.radarbase.appserver.entity.Notification;
+import org.radarbase.appserver.service.FcmDataMessageService;
+import org.radarbase.appserver.service.FcmNotificationService;
+import org.radarbase.appserver.service.UserService;
 import org.radarbase.appserver.service.scheduler.quartz.SchedulerService;
 import org.radarbase.fcm.downstream.FcmSender;
 import org.radarbase.fcm.model.FcmNotificationMessage;
@@ -47,8 +50,11 @@ public class NotificationSchedulerService extends MessageSchedulerService<Notifi
 
     public NotificationSchedulerService(
             @Autowired @Qualifier("fcmSenderProps") FcmSender fcmSender,
-            @Autowired SchedulerService schedulerService) {
-        super(fcmSender, schedulerService);
+            @Autowired SchedulerService schedulerService,
+            @Autowired UserService userService,
+            @Autowired FcmNotificationService notificationService,
+            @Autowired FcmDataMessageService dataMessageService) {
+        super(fcmSender, schedulerService, userService, notificationService, dataMessageService);
     }
 
     private static Map<String, Object> getNotificationMap(Notification notification) {
@@ -98,8 +104,8 @@ public class NotificationSchedulerService extends MessageSchedulerService<Notifi
         } catch (FirebaseMessagingException exc) {
             log.error("Error occurred when sending downstream message.", exc);
             // TODO: update the notification status using event
-            handleErrorCode(exc.getErrorCode());
-            handleFCMErrorCode(exc.getMessagingErrorCode());
+            handleErrorCode(exc.getErrorCode(), notification);
+            handleFCMErrorCode(exc.getMessagingErrorCode(), notification);
         }
     }
 }

--- a/src/main/java/org/radarbase/appserver/service/scheduler/quartz/MessageJob.java
+++ b/src/main/java/org/radarbase/appserver/service/scheduler/quartz/MessageJob.java
@@ -164,13 +164,13 @@ public class MessageJob implements Job {
         log.warn("The FCM service is unavailable.");
         break;
       case UNREGISTERED:
-        FcmUserDto userDto = new FcmUserDto(message.getUser()).setFcmToken("undefined");
+        FcmUserDto userDto = new FcmUserDto(message.getUser());
         log.warn("The Device for user {} was unregistered.", userDto.getSubjectId());
         notificationService.removeNotificationsForUser(
             userDto.getProjectId(), userDto.getSubjectId());
         dataMessageService.removeDataMessagesForUser(
             userDto.getProjectId(), userDto.getSubjectId());
-        userService.updateUser(userDto);
+        userService.checkFcmTokenExistsAndReplace(userDto);
         break;
     }
   }

--- a/src/main/java/org/radarbase/appserver/service/scheduler/quartz/MessageJob.java
+++ b/src/main/java/org/radarbase/appserver/service/scheduler/quartz/MessageJob.java
@@ -21,15 +21,22 @@
 
 package org.radarbase.appserver.service.scheduler.quartz;
 
+import com.google.firebase.ErrorCode;
+import com.google.firebase.messaging.FirebaseMessagingException;
+import com.google.firebase.messaging.MessagingErrorCode;
+import lombok.extern.slf4j.Slf4j;
 import org.quartz.Job;
 import org.quartz.JobDataMap;
 import org.quartz.JobExecutionContext;
 import org.quartz.JobExecutionException;
+import org.radarbase.appserver.dto.fcm.FcmUserDto;
 import org.radarbase.appserver.entity.DataMessage;
+import org.radarbase.appserver.entity.Message;
 import org.radarbase.appserver.entity.Notification;
 import org.radarbase.appserver.service.FcmDataMessageService;
 import org.radarbase.appserver.service.FcmNotificationService;
 import org.radarbase.appserver.service.MessageType;
+import org.radarbase.appserver.service.UserService;
 import org.radarbase.appserver.service.scheduler.DataMessageSchedulerService;
 import org.radarbase.appserver.service.scheduler.NotificationSchedulerService;
 
@@ -40,67 +47,131 @@ import org.radarbase.appserver.service.scheduler.NotificationSchedulerService;
  * @see NotificationSchedulerService
  * @see SchedulerServiceImpl
  */
+@Slf4j
 public class MessageJob implements Job {
 
-    private final transient NotificationSchedulerService notificationSchedulerService;
+  private final transient NotificationSchedulerService notificationSchedulerService;
 
-    private final transient DataMessageSchedulerService dataMessageSchedulerService;
+  private final transient DataMessageSchedulerService dataMessageSchedulerService;
 
-    private final transient FcmNotificationService notificationService;
+  private final transient FcmNotificationService notificationService;
 
-    private final transient FcmDataMessageService dataMessageService;
+  private final transient FcmDataMessageService dataMessageService;
 
-    public MessageJob(
-        NotificationSchedulerService notificationSchedulerService,
-        DataMessageSchedulerService dataMessageSchedulerService,
-        FcmNotificationService notificationService,
-        FcmDataMessageService dataMessageService) {
-      this.notificationSchedulerService = notificationSchedulerService;
-      this.dataMessageSchedulerService = dataMessageSchedulerService;
-      this.notificationService = notificationService;
-      this.dataMessageService = dataMessageService;
-    }
+  private final transient UserService userService;
+
+  public MessageJob(
+      NotificationSchedulerService notificationSchedulerService,
+      DataMessageSchedulerService dataMessageSchedulerService,
+      FcmNotificationService notificationService,
+      FcmDataMessageService dataMessageService,
+      UserService userService) {
+    this.notificationSchedulerService = notificationSchedulerService;
+    this.dataMessageSchedulerService = dataMessageSchedulerService;
+    this.notificationService = notificationService;
+    this.dataMessageService = dataMessageService;
+    this.userService = userService;
+  }
 
   /**
-     * Called by the <code>{@link org.quartz.Scheduler}</code> when a <code>{@link org.quartz.Trigger}
-     * </code> fires that is associated with the <code>Job</code>.
-     *
-     * <p>The implementation may wish to set a {@link JobExecutionContext#setResult(Object) result}
-     * object on the {@link JobExecutionContext} before this method exits. The result itself is
-     * meaningless to Quartz, but may be informative to <code>{@link org.quartz.JobListener}s</code>
-     * or <code>{@link org.quartz.TriggerListener}s</code> that are watching the job's execution.
-     *
-     * @param context context containing jobs details and data added when creating the job.
-     * @throws RuntimeException if there is an CustomExceptionHandler while executing the job.
-     */
-    @Override
-    @SuppressWarnings("PMD.DataflowAnomalyAnalysis")
-    public void execute(JobExecutionContext context) throws JobExecutionException {
-        try {
-            JobDataMap jobDataMap = context.getJobDetail().getJobDataMap();
-            MessageType type = MessageType.valueOf(jobDataMap.getString("messageType"));
-            String projectId = jobDataMap.getString("projectId");
-            String subjectId = jobDataMap.getString("subjectId");
-            Long messageId = jobDataMap.getLong("messageId");
-
-            switch (type) {
-                case NOTIFICATION:
-                    Notification notification =
-                            notificationService.getNotificationByProjectIdAndSubjectIdAndNotificationId(
-                                    projectId, subjectId, messageId);
-                    notificationSchedulerService.send(notification);
-                    break;
-                case DATA:
-                    DataMessage dataMessage =
-                            dataMessageService.getDataMessageByProjectIdAndSubjectIdAndDataMessageId(
-                                    projectId, subjectId, messageId);
-                    dataMessageSchedulerService.send(dataMessage);
-                    break;
-                default:
-                    break;
-            }
-        } catch (Exception e) {
-            throw new JobExecutionException(e);
-        }
+   * Called by the <code>{@link org.quartz.Scheduler}</code> when a <code>{@link org.quartz.Trigger}
+   * </code> fires that is associated with the <code>Job</code>.
+   *
+   * <p>The implementation may wish to set a {@link JobExecutionContext#setResult(Object) result}
+   * object on the {@link JobExecutionContext} before this method exits. The result itself is
+   * meaningless to Quartz, but may be informative to <code>{@link org.quartz.JobListener}s</code>
+   * or <code>{@link org.quartz.TriggerListener}s</code> that are watching the job's execution.
+   *
+   * @param context context containing jobs details and data added when creating the job.
+   * @throws RuntimeException if there is an CustomExceptionHandler while executing the job.
+   */
+  @Override
+  @SuppressWarnings("PMD.DataflowAnomalyAnalysis")
+  public void execute(JobExecutionContext context) throws JobExecutionException {
+    Message message = null;
+    try {
+      JobDataMap jobDataMap = context.getJobDetail().getJobDataMap();
+      MessageType type = MessageType.valueOf(jobDataMap.getString("messageType"));
+      String projectId = jobDataMap.getString("projectId");
+      String subjectId = jobDataMap.getString("subjectId");
+      Long messageId = jobDataMap.getLong("messageId");
+      switch (type) {
+        case NOTIFICATION:
+          Notification notification =
+              notificationService.getNotificationByProjectIdAndSubjectIdAndNotificationId(
+                  projectId, subjectId, messageId);
+          message = notification;
+          notificationSchedulerService.send(notification);
+          break;
+        case DATA:
+          DataMessage dataMessage =
+              dataMessageService.getDataMessageByProjectIdAndSubjectIdAndDataMessageId(
+                  projectId, subjectId, messageId);
+          message = dataMessage;
+          dataMessageSchedulerService.send(dataMessage);
+          break;
+        default:
+          break;
+      }
+    } catch (FirebaseMessagingException exc) {
+      log.error("Error occurred when sending downstream message.", exc);
+      // TODO: update the data message status using event
+      if (message != null) {
+        handleErrorCode(exc.getErrorCode(), message);
+        handleFCMErrorCode(exc.getMessagingErrorCode(), message);
+      }
+    } catch (Exception e) {
+      throw new JobExecutionException(e);
     }
+  }
+
+  protected void handleErrorCode(ErrorCode errorCode, Message message) {
+    // More info on ErrorCode: https://firebase.google.com/docs/reference/fcm/rest/v1/ErrorCode
+    switch (errorCode) {
+      case INVALID_ARGUMENT:
+      case INTERNAL:
+      case ABORTED:
+      case CONFLICT:
+      case CANCELLED:
+      case DATA_LOSS:
+      case NOT_FOUND:
+      case OUT_OF_RANGE:
+      case ALREADY_EXISTS:
+      case DEADLINE_EXCEEDED:
+      case PERMISSION_DENIED:
+      case RESOURCE_EXHAUSTED:
+      case FAILED_PRECONDITION:
+      case UNAUTHENTICATED:
+      case UNKNOWN:
+        break;
+      case UNAVAILABLE:
+        // TODO: Could schedule for retry.
+        log.warn("The FCM service is unavailable.");
+        break;
+    }
+  }
+
+  protected void handleFCMErrorCode(MessagingErrorCode errorCode, Message message) {
+    switch (errorCode) {
+      case INTERNAL:
+      case QUOTA_EXCEEDED:
+      case INVALID_ARGUMENT:
+      case SENDER_ID_MISMATCH:
+      case THIRD_PARTY_AUTH_ERROR:
+        break;
+      case UNAVAILABLE:
+        // TODO: Could schedule for retry.
+        log.warn("The FCM service is unavailable.");
+        break;
+      case UNREGISTERED:
+        FcmUserDto userDto = new FcmUserDto(message.getUser()).setFcmToken("undefined");
+        log.warn("The Device for user {} was unregistered.", userDto.getSubjectId());
+        notificationService.removeNotificationsForUser(
+            userDto.getProjectId(), userDto.getSubjectId());
+        dataMessageService.removeDataMessagesForUser(
+            userDto.getProjectId(), userDto.getSubjectId());
+        userService.updateUser(userDto);
+        break;
+    }
+  }
 }

--- a/src/test/java/org/radarbase/appserver/service/scheduler/NotificationSchedulerServiceTest.java
+++ b/src/test/java/org/radarbase/appserver/service/scheduler/NotificationSchedulerServiceTest.java
@@ -28,16 +28,14 @@ import static org.junit.jupiter.api.Assertions.assertNull;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.junit.jupiter.api.Assertions.fail;
 import static org.mockito.Mockito.mock;
-import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.NONE;
 import static org.radarbase.appserver.controller.RadarUserControllerTest.TIMEZONE;
+import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.NONE;
 
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
-
 import java.time.Duration;
 import java.time.Instant;
 import java.time.temporal.ChronoUnit;
 import java.util.List;
-
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.runner.RunWith;
@@ -54,7 +52,10 @@ import org.radarbase.appserver.config.SchedulerConfig;
 import org.radarbase.appserver.entity.Notification;
 import org.radarbase.appserver.entity.Project;
 import org.radarbase.appserver.entity.User;
+import org.radarbase.appserver.service.FcmDataMessageService;
+import org.radarbase.appserver.service.FcmNotificationService;
 import org.radarbase.appserver.service.MessageType;
+import org.radarbase.appserver.service.UserService;
 import org.radarbase.appserver.service.scheduler.NotificationSchedulerServiceTest.SchedulerServiceTestConfig;
 import org.radarbase.appserver.service.scheduler.quartz.SchedulerServiceImpl;
 import org.radarbase.fcm.downstream.FcmSender;
@@ -222,7 +223,9 @@ class NotificationSchedulerServiceTest {
 
             // mock FCM as we do not want to connect to the server
             return new NotificationSchedulerService(
-                    mock(FcmSender.class), new SchedulerServiceImpl(scheduler));
+                    mock(FcmSender.class), new SchedulerServiceImpl(scheduler),
+                    mock(UserService.class), mock(FcmNotificationService.class),
+                    mock(FcmDataMessageService.class));
         }
     }
 

--- a/src/test/java/org/radarbase/appserver/service/scheduler/NotificationSchedulerServiceTest.java
+++ b/src/test/java/org/radarbase/appserver/service/scheduler/NotificationSchedulerServiceTest.java
@@ -52,10 +52,7 @@ import org.radarbase.appserver.config.SchedulerConfig;
 import org.radarbase.appserver.entity.Notification;
 import org.radarbase.appserver.entity.Project;
 import org.radarbase.appserver.entity.User;
-import org.radarbase.appserver.service.FcmDataMessageService;
-import org.radarbase.appserver.service.FcmNotificationService;
 import org.radarbase.appserver.service.MessageType;
-import org.radarbase.appserver.service.UserService;
 import org.radarbase.appserver.service.scheduler.NotificationSchedulerServiceTest.SchedulerServiceTestConfig;
 import org.radarbase.appserver.service.scheduler.quartz.SchedulerServiceImpl;
 import org.radarbase.fcm.downstream.FcmSender;
@@ -223,9 +220,7 @@ class NotificationSchedulerServiceTest {
 
             // mock FCM as we do not want to connect to the server
             return new NotificationSchedulerService(
-                    mock(FcmSender.class), new SchedulerServiceImpl(scheduler),
-                    mock(UserService.class), mock(FcmNotificationService.class),
-                    mock(FcmDataMessageService.class));
+                    mock(FcmSender.class), new SchedulerServiceImpl(scheduler));
         }
     }
 

